charmhelpers/contrib/network/ovs/ovn.py

```diff
@@ -0,0 +1,148 @@
+            return (rundir)
+    # Fall back to new default
+    return OVN_RUNDIR
+
+
+def ovn_sysconfdir():
+    """Determine path to OVN configuration.
+
+    Prior to OVN 20.03 the default placement of OVN configuration was together
+    with Open vSwitch, from 20.03 and onwards they are kept in a separate
+    directory.
+
+    :returns: path to OVN sysconfdir
+    :rtype: str
+    """
+    OVN_SYSCONFDIR = '/etc/ovn'
+    OVS_SYSCONFDIR = '/etc/openvswitch'
+    for sysconfdir in OVN_SYSCONFDIR, OVS_SYSCONFDIR:
+        if os.path.exists(sysconfdir):
+            return (sysconfdir)
+    # Fall back to new default
+    return OVN_SYSCONFDIR
+
+
+# TODO: Make use of the new `ovn-appctl` if/when that makes sense
+def ovs_appctl(target, *args):
+    """Run `ovs-appctl` for target with args and return output.
+
+    :param target: Name of daemon to contact.  Unless target begins with '/',
+                   `ovs-appctl` looks for a pidfile and will build the path to
+                   a /var/run/openvswitch/target.pid.ctl for you.
+    :type target: str
+    :param args: Command and arguments to pass to `ovs-appctl`
+    :type args: Tuple[str, ...]
+    :returns: Output from command
+    :rtype: str
+    :raises: subprocess.CalledProcessError
+    """
+    # NOTE(fnordahl): The ovsdb-server processes for the OVN databases use a
+    # non-standard naming scheme for their daemon control socket and we need
+    # to pass the full path to the socket.
+    if target in ('ovnnb_db', 'ovnsb_db',):
+        target = os.path.join(ovn_rundir(), target + '.ctl')
+    return utils._run('ovs-appctl', '-t', target, *args)
+
+
+def cluster_status(target, schema=None):
+    """Retrieve status information from clustered OVSDB.
+
+    :param target: Usually one of 'ovsdb-server', 'ovnnb_db', 'ovnsb_db', can
+                   also be full path to control socket.
+    :type target: str
+    :param schema: Database schema name, deduced from target if not provided
+    :type schema: Optional[str]
+    :returns: Structured cluster status data
+    :rtype: Dict[str, Union[str, List[str], Tuple[str, str]]]
+    :raises: subprocess.CalledProcessError
+    """
+    schema_map = {
+        'ovnnb_db': 'OVN_Northbound',
+        'ovnsb_db': 'OVN_Southbound',
+    }
+    status = {}
+    k = ''
+    for line in ovs_appctl(
+            target,
+            'cluster/status',
+            schema or schema_map.get(target)).splitlines():
+        if k and line.startswith(' '):
+            status[k].append(line.lstrip())
+        elif ':' in line:
+            k, v = line.split(':', 1)
+            k = k.lower()
+            k = k.replace(' ', '_')
+            if v:
+                if k in ('cluster_id', 'server_id',):
+                    v = v.replace('(', '')
+                    v = v.replace(')', '')
+                    status[k] = tuple(v.split())
+                else:
+                    status[k] = v.lstrip()
+            else:
+                status[k] = []
```

Comment:
The middle `status[k]` assigns a`str` whereas the other two assign a `List[str]`.  Is that a bug, or by design?

Reply:
This is by design.  The output contain some fields that are multi-line/multi-value which are encoded in an array.  Single-line multi-value fields are encoded in a tuple and single-line single-value fields are plain strings

I'll add a couple of comments to make it more clear what's going on.